[SPARK-32234][SQL] Spark sql commands are failing on selecting the orc tables

SaurabhChawla100 opened the pull request:
### What changes were proposed in this pull request?
Spark sql commands are failing on selecting the orc tables 
Steps to reproduce 
Example 1 - 
Prerequisite -  This is the location(/Users/test/tpcds_scale5data/date_dim) for orc data which is generated by the hive.
```
val table = """CREATE TABLE `date_dim` (
  `d_date_sk` INT,
  `d_date_id` STRING,
  `d_date` TIMESTAMP,
  `d_month_seq` INT,
  `d_week_seq` INT,
  `d_quarter_seq` INT,
  `d_year` INT,
  `d_dow` INT,
  `d_moy` INT,
  `d_dom` INT,
  `d_qoy` INT,
  `d_fy_year` INT,
  `d_fy_quarter_seq` INT,
  `d_fy_week_seq` INT,
  `d_day_name` STRING,
  `d_quarter_name` STRING,
  `d_holiday` STRING,
  `d_weekend` STRING,
  `d_following_holiday` STRING,
  `d_first_dom` INT,
  `d_last_dom` INT,
  `d_same_day_ly` INT,
  `d_same_day_lq` INT,
  `d_current_day` STRING,
  `d_current_week` STRING,
  `d_current_month` STRING,
  `d_current_quarter` STRING,
  `d_current_year` STRING)
USING orc
LOCATION '/Users/test/tpcds_scale5data/date_dim'"""

spark.sql(table).collect

val u = """select date_dim.d_date_id from date_dim limit 5"""

spark.sql(u).collect
```
Example 2

```
  val table = """CREATE TABLE `test_orc_data` (
  `_col1` INT,
  `_col2` STRING,
  `_col3` INT)
  USING orc"""

spark.sql(table).collect

spark.sql("insert into test_orc_data values(13, '155', 2020)").collect

val df = """select _col2 from test_orc_data limit 5"""
spark.sql(df).collect

```



Its Failing with below error 
```
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 2.0 failed 1 times, most recent failure: Lost task 0.0 in stage 2.0 (TID 2, 192.168.0.103, executor driver): java.lang.ArrayIndexOutOfBoundsException: 1
    at org.apache.spark.sql.execution.datasources.orc.OrcColumnarBatchReader.initBatch(OrcColumnarBatchReader.java:156)
    at org.apache.spark.sql.execution.datasources.orc.OrcFileFormat.$anonfun$buildReaderWithPartitionValues$7(OrcFileFormat.scala:258)
    at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.org$apache$spark$sql$execution$datasources$FileScanRDD$$anon$$readCurrentFile(FileScanRDD.scala:141)
    at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:203)
    at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:116)
    at org.apache.spark.sql.execution.FileSourceScanExec$$anon$1.hasNext(DataSourceScanExec.scala:620)
    at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.columnartorow_nextBatch_0$(Unknown Source)
    at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
    at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
    at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:729)
    at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:343)
    at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:895)
    at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:895)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:372)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:336)
    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
    at org.apache.spark.scheduler.Task.run(Task.scala:133)
    at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:445)
    at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1489)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:448)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)`
```

The reason behind this initBatch is not getting the schema that is needed to find out the column value in OrcFileFormat.scala
```
batchReader.initBatch(
 TypeDescription.fromString(resultSchemaString)
```

### Why are the changes needed?
Spark sql queries for orc tables are failing


### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Unit test is added for this .Also Tested through spark shell and spark submit the failing queries

SaurabhChawla100 committed:
add the code change to fix the sql query

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun reviewed (changes requested):
Thank you for your contribution, @SaurabhChawla100 . 
In order to prevent the future regression, could you make a UT with your example, please?

dongjoon-hyun commented:
ok to test

SparkQA commented:
**[Test build #125398 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125398/testReport)** for PR 29045 at commit [`8e5b79e`](https://github.com/apache/spark/commit/8e5b79ee836caf4df3fff7c013e8189b96f35424).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30011/
Test PASSed.

SparkQA commented:
**[Test build #125398 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125398/testReport)** for PR 29045 at commit [`8e5b79e`](https://github.com/apache/spark/commit/8e5b79ee836caf4df3fff7c013e8189b96f35424).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125398/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #125411 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125411/testReport)** for PR 29045 at commit [`8e5b79e`](https://github.com/apache/spark/commit/8e5b79ee836caf4df3fff7c013e8189b96f35424).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30022/
Test PASSed.

SparkQA commented:
**[Test build #125411 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125411/testReport)** for PR 29045 at commit [`8e5b79e`](https://github.com/apache/spark/commit/8e5b79ee836caf4df3fff7c013e8189b96f35424).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125411/
Test FAILed.

SaurabhChawla100 commented:
> Thank you for your contribution, @SaurabhChawla100 .
> In order to prevent the future regression, could you make a UT with your example, please?

Sure I will add the unit test

SaurabhChawla100 committed:
handle the scenrios for orc file created by hive data

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30115/
Test PASSed.

SparkQA commented:
**[Test build #125502 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125502/testReport)** for PR 29045 at commit [`75e8833`](https://github.com/apache/spark/commit/75e8833a4c5c1b9cf48c4bc3322cd8143a042b38).

SparkQA commented:
**[Test build #125502 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125502/testReport)** for PR 29045 at commit [`75e8833`](https://github.com/apache/spark/commit/75e8833a4c5c1b9cf48c4bc3322cd8143a042b38).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125502/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #125520 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125520/testReport)** for PR 29045 at commit [`75e8833`](https://github.com/apache/spark/commit/75e8833a4c5c1b9cf48c4bc3322cd8143a042b38).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30135/
Test PASSed.

SparkQA commented:
**[Test build #125520 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125520/testReport)** for PR 29045 at commit [`75e8833`](https://github.com/apache/spark/commit/75e8833a4c5c1b9cf48c4bc3322cd8143a042b38).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125520/
Test FAILed.

maropu commented:
retest this please

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReaderSuite.scala:
```diff
@@ -77,4 +77,43 @@ class OrcColumnarBatchReaderSuite extends QueryTest with SharedSparkSession {
+    assert(error == null)
+    spark.sql(s"DROP TABLE IF EXISTS test_date_spark_orc")
+  }
+
```
nit: remove this blank.

SaurabhChawla100 replied:
done

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReaderSuite.scala:
```diff
@@ -77,4 +77,43 @@ class OrcColumnarBatchReaderSuite extends QueryTest with SharedSparkSession {
     }
   }
+
+  test("orc data created by the hive tables having _col fields name") {
```
Plz add the prefix `test("SPARK-32234: orc data...`.

SaurabhChawla100 replied:
done

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReaderSuite.scala:
```diff
@@ -77,4 +77,43 @@ class OrcColumnarBatchReaderSuite extends QueryTest with SharedSparkSession {
+    spark.sql(s"DROP TABLE IF EXISTS test_date_hive_orc")
+  }
+
+  test("orc data created by the spark having proper fields name") {
```
ditto

SaurabhChawla100 replied:
done

SparkQA commented:
**[Test build #125555 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125555/testReport)** for PR 29045 at commit [`75e8833`](https://github.com/apache/spark/commit/75e8833a4c5c1b9cf48c4bc3322cd8143a042b38).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30170/
Test PASSed.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReaderSuite.scala:
```diff
@@ -77,4 +77,43 @@ class OrcColumnarBatchReaderSuite extends QueryTest with SharedSparkSession {
+        error = e
+    }
+    assert(error == null)
+    spark.sql(s"DROP TABLE IF EXISTS test_date_hive_orc")
```
Please check the other tests carefully then follow how to write tests there. How about refactoring it like this?
```
    withTable("test_date_hive_orc") {
      spark.sql(
        s"""
           |CREATE TABLE test_date_hive_orc
           |  (col1 INT, col2 STRING, col3 INT)
           |  USING orc
         """.stripMargin)
      spark.sql(
        s"""
           |INSERT INTO test_date_hive_orc VALUES
           |  (9, '12', 2020)
         """.stripMargin)

      val df = spark.sql("SELECT col2 FROM test_date_hive_orc")
      checkAnswer(df, Row(...))
    }
```

SaurabhChawla100 replied:
Refactored the unit test

SaurabhChawla100 committed:
Refactor the unit test as per the review comments

SparkQA commented:
**[Test build #125555 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125555/testReport)** for PR 29045 at commit [`75e8833`](https://github.com/apache/spark/commit/75e8833a4c5c1b9cf48c4bc3322cd8143a042b38).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125555/
Test FAILed.

SparkQA commented:
**[Test build #125593 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125593/testReport)** for PR 29045 at commit [`5b8edfb`](https://github.com/apache/spark/commit/5b8edfbd48a70e72960e6badb658e9ffbfeb7577).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30209/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125593/
Test FAILed.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReaderSuite.scala:
```diff
@@ -77,4 +77,64 @@ class OrcColumnarBatchReaderSuite extends QueryTest with SharedSparkSession {
+      assert(error == null)
+      spark.sql(
+        s"""
+           |DROP TABLE IF
```
you don't need to drop the table. Please see the implementation of `withTable`.

SaurabhChawla100 replied:
yes in withTable table its handled, Removed this drop table .

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReaderSuite.scala:
```diff
@@ -77,4 +77,64 @@ class OrcColumnarBatchReaderSuite extends QueryTest with SharedSparkSession {
+        case e: Throwable =>
+          error = e
+      }
+      assert(error == null)
```
I think you don't need this check.

SaurabhChawla100 replied:
Yes not required already handled in the framework in checkAnswer

SaurabhChawla100 committed:
Remove drop table from the test as its already handled in withTable

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30280/
Test PASSed.

SparkQA commented:
**[Test build #125668 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125668/testReport)** for PR 29045 at commit [`c555f3c`](https://github.com/apache/spark/commit/c555f3c8658ad169d56d26249628049d1d01cdd6).

SparkQA commented:
**[Test build #125668 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125668/testReport)** for PR 29045 at commit [`c555f3c`](https://github.com/apache/spark/commit/c555f3c8658ad169d56d26249628049d1d01cdd6).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125668/
Test FAILed.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReaderSuite.scala:
```diff
@@ -77,4 +77,43 @@ class OrcColumnarBatchReaderSuite extends QueryTest with SharedSparkSession {
   }
+
+  test("SPARK-32234: orc data created by the hive tables having _col fields name") {
+    var error: Throwable = null
```
please remove this.

SaurabhChawla100 replied:
done

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReaderSuite.scala:
```diff
@@ -77,4 +77,43 @@ class OrcColumnarBatchReaderSuite extends QueryTest with SharedSparkSession {
+    withTable("test_date_hive_orc") {
+      spark.sql(
+        """
+          |CREATE TABLE `test_date_hive_orc`
```
nit: we need the backquotes?

SaurabhChawla100 replied:
not required

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReaderSuite.scala:
```diff
@@ -77,4 +77,43 @@ class OrcColumnarBatchReaderSuite extends QueryTest with SharedSparkSession {
+          |  values(9, '12', 2020)
+        """.stripMargin)
+
+      val df = spark.sql("select d_date_id from test_date_spark_orc")
```
nit: please use the uppercases for SQL keywrods (e.g., SELECT) where possible

SaurabhChawla100 replied:
Done

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala:
```diff
@@ -181,10 +183,19 @@ class OrcFileFormat
+          // in the physical schema, there is a need to send the
+          // entire dataSchema instead of required schema
+          val orcFieldNames = reader.getSchema.getFieldNames.asScala
+          if (orcFieldNames.forall(_.startsWith("_col"))) {
```
What does this code mean? `_col` needs to be hard-coded?

SaurabhChawla100 replied:
So this is for a ORC file written by Hive, no field names in the physical schema. In that case it its having names  like _col1, _col2 etc.

Check this code for reference https://github.com/apache/spark/blob/84db660ebef4f9c543ab2709103c4542b407a829/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala#L133

SaurabhChawla100 committed:
Refactor the code change as per the review comments

SparkQA commented:
**[Test build #125675 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125675/testReport)** for PR 29045 at commit [`95ca209`](https://github.com/apache/spark/commit/95ca209ce1e47be0ed7c80ca6ba3895bae61c7d0).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30288/
Test PASSed.

SparkQA commented:
**[Test build #125675 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125675/testReport)** for PR 29045 at commit [`95ca209`](https://github.com/apache/spark/commit/95ca209ce1e47be0ed7c80ca6ba3895bae61c7d0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125675/
Test PASSed.

dongjoon-hyun reviewed (commented):
Thank you for updating, but could you update the PR description with the reproducible example? If someone is following the example, it will not fail because they don't have a data in `/Users/test/tpcds_scale5data/date_dim`. Also, please remove irrelevant stuff like `TBLPROPERTIES`.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReaderSuite.scala:
```diff
@@ -77,4 +77,44 @@ class OrcColumnarBatchReaderSuite extends QueryTest with SharedSparkSession {
+    }
+  }
+
+  test("SPARK-32234: orc data created by the spark having proper fields name") {
```
Shall we remove this test case because this test case pass without your patch?

SaurabhChawla100 replied:
done

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala:
```diff
@@ -181,10 +183,19 @@ class OrcFileFormat
+          val orcFieldNames = reader.getSchema.getFieldNames.asScala
+          if (orcFieldNames.forall(_.startsWith("_col"))) {
+            resultSchemaString = OrcUtils.orcTypeDescriptionString(actualSchema)
+          }
```
Can we have the above logic inside `OrcUtils.requestedColumnIds` function instead of this file, @SaurabhChawla100 ?

SaurabhChawla100 replied:
I have refactored the code an include this in OrcUtils.requestedColumnIds

dongjoon-hyun reviewed (changes requested):
@SaurabhChawla100 . Thanks for updating. Apache Spark has two ORC implementations `native` and `hive`. Could you add the following test coverage more?

1. A test coverage for `hive` ORC implementation
- sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala

2. A non-vectorized code test coverage for `native` ORC implementaion
- `spark.sql.orc.enableVectorizedReader=false`

SaurabhChawla100 committed:
add the unit test for ORC_VECTORIZED_READER_ENABLED

SaurabhChawla100 committed:
add the code change to handle the schema sent through requestedColumnIds

SaurabhChawla100 commented:
> @SaurabhChawla100 . Thanks for updating. Apache Spark has two ORC implementations `native` and `hive`. Could you add the following test coverage more?
> 
> 1. A test coverage for `hive` ORC implementation
> 
> * sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala
> 
> 1. A non-vectorized code test coverage for `native` ORC implementaion
> 
> * `spark.sql.orc.enableVectorizedReader=false`

Added both the requested unit test

SparkQA commented:
**[Test build #125709 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125709/testReport)** for PR 29045 at commit [`52066be`](https://github.com/apache/spark/commit/52066beaa61f13a3acf42b12b5df7fac7f9b87a0).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30322/
Test PASSed.

SaurabhChawla100 commented:
> Thank you for updating, but could you update the PR description with the reproducible example? If someone is following the example, it will not fail because they don't have a data in `/Users/test/tpcds_scale5data/date_dim`. Also, please remove irrelevant stuff like `TBLPROPERTIES`.

@dongjoon-hyun  - Thank you for reviewing this PR . 
I have updated the PR description with prerequisite create date from Hive  '/Users/test/tpcds_scale5data/date_dim`. This is needed since as that is the real time scenario where its failing , Also I have added the second example in the PR description where this error can be reproduce.

SaurabhChawla100 committed:
removed the extra import from OrcFileFormat.scala

SparkQA commented:
**[Test build #125710 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125710/testReport)** for PR 29045 at commit [`46c65a1`](https://github.com/apache/spark/commit/46c65a100331b5c908db71424ea00cdfd3887535).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30323/
Test PASSed.

SparkQA commented:
**[Test build #125709 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125709/testReport)** for PR 29045 at commit [`52066be`](https://github.com/apache/spark/commit/52066beaa61f13a3acf42b12b5df7fac7f9b87a0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125709/
Test PASSed.

SparkQA commented:
**[Test build #125710 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125710/testReport)** for PR 29045 at commit [`46c65a1`](https://github.com/apache/spark/commit/46c65a100331b5c908db71424ea00cdfd3887535).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125710/
Test PASSed.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala:
```diff
@@ -185,10 +185,15 @@ class OrcFileFormat
+      }
+      OrcConf.MAPRED_INPUT_SCHEMA.setString(conf, resultSchemaString)
+
+      if (requestedColIdsOrEmptyFile._1.isEmpty) {
```
Thank you for updating, @SaurabhChawla100 . However, `_1` and `_2` is not good in general. You may try like the following at line 182.
```scala
- val requestedColIdsOrEmptyFile =
+ val (requestedColIdsOrEmptyFile, sendActualSchema) =
```

SaurabhChawla100 replied:
done

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcPartitionReaderFactory.scala:
```diff
@@ -80,10 +80,10 @@ case class OrcPartitionReaderFactory(
       }
 
-    if (requestedColIdsOrEmptyFile.isEmpty) {
+    if (requestedColIdsOrEmptyFile._1.isEmpty) {
```
ditt.

Please try the following at line 77.
```scala
- val requestedColIdsOrEmptyFile =
+ val (requestedColIdsOrEmptyFile, _) =
```

SaurabhChawla100 replied:
done

dongjoon-hyun commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala:
```diff
@@ -288,4 +288,56 @@ class HiveOrcQuerySuite extends OrcQueryTest with TestHiveSingleton {
+    }
+  }
+
+  test("SPARK-32234: orc data created by the hive tables having _col fields name" +
```
Please merge two test cases like the following in order to have a full test coverage.

```scala
Seq("native", "hive").foreach { orcImpl =>
  Seq(false, true).foreach { vectorized =>
```

SaurabhChawla100 replied:
Merge the test

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:
```diff
@@ -116,47 +116,53 @@ object OrcUtils extends Logging {
-   * requested column doesn't exist in the ORC file. Returns None if the given ORC file is empty.
+   * Returns the requested column ids from the given ORC file and Boolean flag to use actual
+   * schema or result schema. Column id can be -1, which means the requested column doesn't
+   * exist in the ORC file. Returns None if the given ORC file is empty.
```
Could you add `@return` for describing what's a return value `(Option[Array[Int]], Boolean)`?

SaurabhChawla100 replied:
done

SaurabhChawla100 committed:
refactor the code using seprate varible for sendActualSchema

SaurabhChawla100 committed:
Merge the orcimpl and vectorised test for complete code coverage

SparkQA commented:
**[Test build #125750 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125750/testReport)** for PR 29045 at commit [`5b93849`](https://github.com/apache/spark/commit/5b9384938d9570f5b8982a6e713529f0c14d6222).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30361/
Test PASSed.

SparkQA commented:
**[Test build #125752 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125752/testReport)** for PR 29045 at commit [`3c4baeb`](https://github.com/apache/spark/commit/3c4baeb5a3ede6559273a503193bd03a8ad5aa41).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30363/
Test PASSed.

SparkQA commented:
**[Test build #125750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125750/testReport)** for PR 29045 at commit [`5b93849`](https://github.com/apache/spark/commit/5b9384938d9570f5b8982a6e713529f0c14d6222).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #125752 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125752/testReport)** for PR 29045 at commit [`3c4baeb`](https://github.com/apache/spark/commit/3c4baeb5a3ede6559273a503193bd03a8ad5aa41).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125752/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125750/
Test FAILed.

SparkQA commented:
**[Test build #125760 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125760/testReport)** for PR 29045 at commit [`4469e6d`](https://github.com/apache/spark/commit/4469e6d1c5970bc5de3d4a4dc728e9775e43e141).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30371/
Test PASSed.

SaurabhChawla100 commented:
@dongjoon-hyun / @maropu  - Its seems like some issue happened after I rebase this current branch with the master branch. 

Its now showing number of files change is -185. Not sure why its showing this difference

SaurabhChawla100 committed:
Merge remote-tracking branch 'os/master' into SPARK-32234

SaurabhChawla100 commented:
> @dongjoon-hyun / @maropu - Its seems like some issue happened after I rebase this current branch with the master branch.
> 
> Its now showing number of files change is -185. Not sure why its showing this difference

Now its fine

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30375/
Test PASSed.

SparkQA commented:
**[Test build #125764 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125764/testReport)** for PR 29045 at commit [`743ffe3`](https://github.com/apache/spark/commit/743ffe3775c03ef1b6a37b48cc3f3fe767f64a52).

SparkQA commented:
**[Test build #125760 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125760/testReport)** for PR 29045 at commit [`4469e6d`](https://github.com/apache/spark/commit/4469e6d1c5970bc5de3d4a4dc728e9775e43e141).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125760/
Test PASSed.

SparkQA commented:
**[Test build #125764 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125764/testReport)** for PR 29045 at commit [`743ffe3`](https://github.com/apache/spark/commit/743ffe3775c03ef1b6a37b48cc3f3fe767f64a52).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait BaseScriptTransformationExec extends UnaryExecNode `
  * `abstract class BaseScriptTransformationWriterThread(`
  * `abstract class BaseScriptTransformIOSchema extends Serializable `
  * `case class HiveScriptTransformationExec(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125764/
Test PASSed.

gatorsmile commented:
cc @gengliangwang @cloud-fan @MaxGekk

cloud-fan commented:
> The reason behind this initBatch is not getting the schema that is needed to find out the column value in OrcFileFormat.scala

Can you be more specific about the problem? Are you saying that the actual file schema doesn't match the table schema specified by the user?

SaurabhChawla100 commented:
> Can you be more specific about the problem? Are you saying that the actual file schema doesn't match the table schema specified by the user?

So in case of orc data created by the hive no field names in the physical schema. Please find the below code for reference.
https://github.com/apache/spark/blob/24be81689cee76e03cd5136dfd089123bbff4595/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala#L133

So from this code we are sending the index of the col from the dataschema.

But Where as in the below code , we are passing the input result schema and that result schema will not have that index number that is passed from OrcUtils.scala
https://github.com/apache/spark/blob/24be81689cee76e03cd5136dfd089123bbff4595/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala#L211

For example - 

```
val u = """select date_dim.d_date_id from date_dim limit 5"""

spark.sql(u).collect
```

Here the value of index(d_date_id) returned by the OrcUtils.scala#L133 is 2 

where the resultSchema passed in OrcFileFormat.scala#L211 is having only one  struct<`d_date_id`:string>

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala:
```diff
@@ -179,12 +179,17 @@ class OrcFileFormat
         }
 
+      if (sendActualSchema) {
+        resultSchemaString = OrcUtils.orcTypeDescriptionString(actualSchema)
```
do you mean we can't do column pruning in this case?

SaurabhChawla100 replied:
Yes for this case we cannot do it ,so as the result of which we have the give the actual schema

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:
```diff
@@ -116,47 +116,53 @@ object OrcUtils extends Logging {
-      conf: Configuration): Option[Array[Int]] = {
+      conf: Configuration): (Option[Array[Int]], Boolean) = {
+    var sendActualSchema = false
     val orcFieldNames = reader.getSchema.getFieldNames.asScala
```
let's say we are running the test `CREATE TABLE test_hive_orc_impl ...`, what's the value of `dataSchema`, `requiredSchema` and `orcFieldNames` when we reach here?

SaurabhChawla100 replied:
For this example query
```
val u = """select date_dim.d_year from date_dim limit 5"""

spark.sql(u).collect
```

value of the orcFieldNames
```
val of orcFieldNames = {Wrappers$JListWrapper@13621} "Wrappers$JListWrapper" size = 28
 0 = "_col0"
 1 = "_col1"
 2 = "_col2"
 3 = "_col3"
 4 = "_col4"
 5 = "_col5"
 6 = "_col6"
 7 = "_col7"
 8 = "_col8"
 9 = "_col9"
 10 = "_col10"
```

value of dataSchema
```
value of dataSchema = {StructType@13616} "StructType" size = 28
 0 = {StructField@16487} "StructField(d_date_sk,IntegerType,true)"
 1 = {StructField@16488} "StructField(d_date_id,StringType,true)"
 2 = {StructField@16489} "StructField(d_date,TimestampType,true)"
 3 = {StructField@16490} "StructField(d_month_seq,IntegerType,true)"
 4 = {StructField@16491} "StructField(d_week_seq,IntegerType,true)"
 5 = {StructField@16492} "StructField(d_quarter_seq,IntegerType,true)"
 6 = {StructField@16493} "StructField(d_year,IntegerType,true)"
 7 = {StructField@16494} "StructField(d_dow,IntegerType,true)"
 8 = {StructField@16495} "StructField(d_moy,IntegerType,true)"
 9 = {StructField@16496} "StructField(d_dom,IntegerType,true)"
 10 = {StructField@16497} "StructField(d_qoy,IntegerType,true)"
```

value of requiredSchema

```
result = {StructType@16990} "StructType" size = 1
 0 = {StructField@17128} "StructField(d_year,IntegerType,true)"
```

cloud-fan replied:
Please correct me if I'm wrong:
1. the physical orc file schema is `_col0`, ...
2. the table schema in metastore is `d_date_sk`, ...
3. the query only requires only `d_year`

I don't know why the query fails. The `requestedColumnIds` will be `[6]` and the orc reader will read the `_col6` column. Everything should be fine.

SaurabhChawla100 replied:
In this code , this is the place where we are getting exception

https://github.com/apache/spark/blob/d6a68e0b67ff7de58073c176dd097070e88ac831/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java#L183

So the code is creating a wrap VectorizedRowBatchWrap using the result schema which is 
```
result = {StructType@16990} "StructType" size = 1
 0 = {StructField@17128} "StructField(d_year,IntegerType,true)"
```

https://github.com/apache/spark/blob/d6a68e0b67ff7de58073c176dd097070e88ac831/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java#L149 

Now wrap is created with size 1 

and in this code 
` code orcVectorWrappers[i] = new OrcColumnVector(dt, wrap.batch().cols[colId]);`

value of colId will be 6 (which is the requestedColumnIds will be [6])

Now the size of wrap is 1 and code is requesting the fetch wrap.batch().cols[6] so here its getting ArrayIndexOutOfBoundsException

cloud-fan replied:
Does the failure go away if the physical orc file schema matches the table schema in metastore?

SaurabhChawla100 replied:
yes, if that would be the case then it will be same as how the data created from the spark Application using orc. And it follows the code flow as its working today for spark orc data source tables.

So this only failing for the orc data created by hive, So if I create the data from this using the spark orc datasource , This error is not coming

```
val u = """select * from date_dim limit 5"""

scala> spark.sql(u).write.format("orc").save("/Users/tpcdsdata/testFS/testorc")

val table = """CREATE TABLE `date_dim345` (
     |   `d_date_sk` INT,
     |   `d_date_id` STRING,
     |   `d_date` TIMESTAMP,
     |   `d_month_seq` INT,
     |   `d_week_seq` INT,
     |   `d_quarter_seq` INT,
     |   `d_year` INT,
     |   `d_dow` INT,
     |   `d_moy` INT,
     |   `d_dom` INT,
     |   `d_qoy` INT,
     |   `d_fy_year` INT,
     |   `d_fy_quarter_seq` INT,
     |   `d_fy_week_seq` INT,
     |   `d_day_name` STRING,
     |   `d_quarter_name` STRING,
     |   `d_holiday` STRING,
     |   `d_weekend` STRING,
     |   `d_following_holiday` STRING,
     |   `d_first_dom` INT,
     |   `d_last_dom` INT,
     |   `d_same_day_ly` INT,
     |   `d_same_day_lq` INT,
     |   `d_current_day` STRING,
     |   `d_current_week` STRING,
     |   `d_current_month` STRING,
     |   `d_current_quarter` STRING,
     |   `d_current_year` STRING)
     | USING orc
     | LOCATION '/Users/tpcdsdata/testFS/testorc/'"""

spark.sql(table).collect
val u = """select date_dim.d_date_id from date_dim345 limit 5"""
```
Now this is having the correct value of  physical orc file not _col1, _col2 etc
```
orcFieldNames = {Wrappers$JListWrapper@19940} "Wrappers$JListWrapper" size = 28
 1 = "d_date_id"
 9 = "d_dom"
 7 = "d_dow"
 11 = "d_fy_year"
 5 = "d_quarter_seq"
 0 = "d_date_sk"
 8 = "d_moy"
 10 = "d_qoy"
 4 = "d_week_seq"
 6 = "d_year"
 3 = "d_month_seq"
 2 = "d_date"
```

cloud-fan replied:
following your previous explanation in https://github.com/apache/spark/pull/29045#discussion_r454234413

1. the code is creating a wrap VectorizedRowBatchWrap using the result schema which is `<d_year, int>`
2. `orcVectorWrappers[i] = new OrcColumnVector(dt, wrap.batch().cols[colId]);` accesses the 6th column of the orc batch.

Why it doesn't fail when the physical orc file schema matches the table schema in metastore?

SaurabhChawla100 replied:
In this case code follows this path

https://github.com/apache/spark/blob/d6a68e0b67ff7de58073c176dd097070e88ac831/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala#L158

Now requiredSchema (result schema) is used for getting the column id instead of entire dataschema . So in this scenario required requiredSchema(requiredSchema) is used to create the VectorizedRowBatchWrap
wrap = new VectorizedRowBatchWrap(orcSchema.createRowBatch(capacity)); 


Where as in failure case column id got from the dataschema and VectorizedRowBatchWrap is created using the requiredSchema (result schema).

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala:
```diff
@@ -160,12 +160,12 @@ class OrcFileFormat
     val capacity = sqlConf.orcVectorizedReaderBatchSize
 
-    val resultSchemaString = OrcUtils.orcTypeDescriptionString(resultSchema)
-    OrcConf.MAPRED_INPUT_SCHEMA.setString(hadoopConf, resultSchemaString)
```
After reading the code more, I think this is the real problem. When the physical file schema doesn't match the table schema, this `resultSchemaString` is wrong.

I think we should set this config in the executor side, where we know the physical file schema. The `OrcUtils.requestedColumnIds` should not only report the column indices, but also the actual `requiredSchema`.

SaurabhChawla100 replied:
will this not help 
I have already added this in the code
`OrcConf.MAPRED_INPUT_SCHEMA.setString(conf, resultSchemaString)`

SaurabhChawla100 replied:
@cloud-fan - Do we need to make the similar change in these file OrcPartitionReaderFactory

override def buildReader(file: PartitionedFile): PartitionReader[InternalRow] = {
and 
override def buildColumnarReader(file: PartitionedFile): PartitionReader[ColumnarBatch] = {

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala:
```diff
@@ -288,4 +288,35 @@ class HiveOrcQuerySuite extends OrcQueryTest with TestHiveSingleton {
+              spark.sql(
+                s"""
+                   | CREATE TABLE test_hive_orc_impl
+                   | (_col1 INT, _col2 STRING, _col3 INT)
```
can this test reproduce the bug? the table schema matches the physical file schema.

SaurabhChawla100 replied:
yes this can be reproduce by this also, 

But I have attached the date_dim tpcds orc data in the jira itself

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:
```diff
@@ -116,47 +116,53 @@ object OrcUtils extends Logging {
+      conf: Configuration): (Option[Array[Int]], Boolean) = {
+    var sendActualSchema = false
     val orcFieldNames = reader.getSchema.getFieldNames.asScala
     if (orcFieldNames.isEmpty) {
```
If we can't do column pruning for the _col case, shall we simply return None? e.g.
```
if (orcFieldNames.isEmpty || orcFieldNames.forall(_.startsWith("_col"))) {
  None
}
```

SaurabhChawla100 replied:
This will not give correct result, On changing the above comment, results are showing empty rows

```
scala> val u = """select date_dim.d_date_id from date_dim limit 5"""
u: String = select date_dim.d_date_id from date_dim limit 5

scala> spark.sql(u).collect
res1: Array[org.apache.spark.sql.Row] = Array()

scala> val u = """select * from date_dim limit 5"""
u: String = select * from date_dim limit 5
scala> spark.sql(u).collect
res2: Array[org.apache.spark.sql.Row] = Array()
```

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:
```diff
@@ -116,47 +116,53 @@ object OrcUtils extends Logging {
+            sendActualSchema = true
             index
           } else {
             -1
```
I tried the test locally, and saw warning messages like
```
10:45:52.783 WARN org.apache.orc.impl.SchemaEvolution: Column names are missing from this file. This is caused by a writer earlier than HIVE-4243. The reader will reconcile schemas based on index. File type: struct<_col1:int,_col2:string,_col3:int>, reader type: struct<_col2:string>
{9}
```

I think we can't do column pruning anyway if the physical file schema is `_col0`, ... We can always return true in this branch.

SaurabhChawla100 replied:
for -1 also we need to send true

SaurabhChawla100 replied:
added for -1

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:
```diff
@@ -116,47 +116,53 @@ object OrcUtils extends Logging {
       requiredSchema: StructType,
       reader: Reader,
-      conf: Configuration): Option[Array[Int]] = {
+      conf: Configuration): (Option[Array[Int]], Boolean) = {
```
I think the return type should be `Option[(Array[Int], Boolean)]`

SaurabhChawla100 replied:
I did similar change by returning the one attribute in the return type. I was asked to make it two attributes to return from this method
Please find the review comment for that  - https://github.com/apache/spark/pull/29045#discussion_r453335630

So I change it after this review comment

cloud-fan replied:
how about
```
def requestedColumnIds(...): Option[(Array[Int], Boolean)]
...
val result = ... { OrcUtils.requestedColumnIds... }
if (result.isEmpty) {
  Iterator.empty
} else {
  val (requestedColIds, canPruneCols) = result.get
  ...
}
```

SaurabhChawla100 replied:
done

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala:
```diff
@@ -179,12 +179,17 @@ class OrcFileFormat
       val fs = filePath.getFileSystem(conf)
       val readerOptions = OrcFile.readerOptions(conf).filesystem(fs)
-      val requestedColIdsOrEmptyFile =
+      val (requestedColIdsOrEmptyFile, sendActualSchema) =
```
`sendActualSchema` is a bit confusing, how about `canPruneCols`?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcPartitionReaderFactory.scala:
```diff
@@ -74,7 +74,7 @@ case class OrcPartitionReaderFactory(
     val fs = filePath.getFileSystem(conf)
     val readerOptions = OrcFile.readerOptions(conf).filesystem(fs)
-    val requestedColIdsOrEmptyFile =
+    val (requestedColIdsOrEmptyFile, _) =
```
We should fix the file source v2 code path as well.

SaurabhChawla100 replied:
done

SaurabhChawla100 committed:
reafactor the code and add similar logic in file source v2 code path

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30460/
Test PASSed.

SparkQA commented:
**[Test build #125850 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125850/testReport)** for PR 29045 at commit [`9de3516`](https://github.com/apache/spark/commit/9de35164528dc67a23b5137d0a5e81dd9371d6e4).

SaurabhChawla100 committed:
refactor the code

SparkQA commented:
**[Test build #125851 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125851/testReport)** for PR 29045 at commit [`f8ece1f`](https://github.com/apache/spark/commit/f8ece1fe1eeb9e2cdf2c591dc7e2cfc9af002340).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30461/
Test PASSed.

SparkQA commented:
**[Test build #125850 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125850/testReport)** for PR 29045 at commit [`9de3516`](https://github.com/apache/spark/commit/9de35164528dc67a23b5137d0a5e81dd9371d6e4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125850/
Test PASSed.

SparkQA commented:
**[Test build #125851 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125851/testReport)** for PR 29045 at commit [`f8ece1f`](https://github.com/apache/spark/commit/f8ece1fe1eeb9e2cdf2c591dc7e2cfc9af002340).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125851/
Test FAILed.

SaurabhChawla100 commented:
Retest this please

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala:
```diff
@@ -179,12 +179,17 @@ class OrcFileFormat
             isCaseSensitive, dataSchema, requiredSchema, reader, conf)
         }
 
+      if (!canPruneCols) {
```
nit: we can simplify the code a bit
```
val resultSchemaString = if (canPruneCols) {
  OrcUtils.orcTypeDescriptionString(resultSchema)
} else {
  OrcUtils.orcTypeDescriptionString(StructType(dataSchema.fields ++ partitionSchema.fields))
}
```

Then we don't need to keep the `val actualSchema =...` and `var resultSchemaString =...` at the beginning.

SaurabhChawla100 replied:
done

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:
```diff
@@ -116,47 +116,53 @@ object OrcUtils extends Logging {
       reader: Reader,
-      conf: Configuration): Option[Array[Int]] = {
+      conf: Configuration): (Option[Array[Int]], Boolean) = {
+    var canPruneCols = true
```
do we really need it? We can just use boolean literal in the places that return the value.

SaurabhChawla100 replied:
Removed it

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala:
```diff
@@ -288,4 +288,35 @@ class HiveOrcQuerySuite extends OrcQueryTest with TestHiveSingleton {
+        withSQLConf(
+          SQLConf.ORC_IMPLEMENTATION.key -> orcImpl,
+          SQLConf.ORC_VECTORIZED_READER_ENABLED.key -> vectorized) {
+          withTempPath { dir =>
```
nit: we don't need to provide a custom location. CREATE TABLE without LOCATION clause can also reproduce it.

SaurabhChawla100 replied:
Removed it

SaurabhChawla100 committed:
Refactor some method return type and use it the code

SaurabhChawla100 committed:
Removed the code of location for create table in unit test

SaurabhChawla100 committed:
fix the scala style in the unit test

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:
```diff
@@ -116,15 +116,16 @@ object OrcUtils extends Logging {
   /**
-   * Returns the requested column ids from the given ORC file. Column id can be -1, which means the
-   * requested column doesn't exist in the ORC file. Returns None if the given ORC file is empty.
+   * @return Returns the requested column ids from the given ORC file and Boolean flag to use actual
```
can we update the comment a little bit?

SaurabhChawla100 replied:
updated the comment

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcPartitionReaderFactory.scala:
```diff
@@ -66,24 +66,28 @@ case class OrcPartitionReaderFactory(
+      } else {
+        OrcUtils.orcTypeDescriptionString(StructType(dataSchema.fields ++ partitionSchema.fields))
+      }
+      OrcConf.MAPRED_INPUT_SCHEMA.setString(conf, resultSchemaString)
```
To avoid duplicated code, we can move these to the `requestedColumnIds` method
```
def requestedColumnIds(..., partitionSchema: StructType): Option[Array[Int]] = {
  ...
  if (orcFieldNames.isEmpty) {
    None
  } else if (orcFieldNames.forall(_.startsWith("_col"))) {
    OrcConf.MAPRED_INPUT_SCHEMA.setString(conf,
      OrcUtils.orcTypeDescriptionString(StructType(dataSchema.fields ++ partitionSchema.fields)))
    ...
  } else {
    OrcConf.MAPRED_INPUT_SCHEMA.setString(conf,
      OrcUtils.orcTypeDescriptionString(StructType(requiredSchema.fields ++ partitionSchema.fields)))
    ...
  }
}
```

SaurabhChawla100 replied:
@cloud-fan - In this we need to return the resultSchemaString from this method  Option[(Array[Int], String)]
 which is for else if (orcFieldNames.forall(_.startsWith("_col"))) {
 val resultSchemaString = OrcUtils.orcTypeDescriptionString(StructType(dataSchema.fields ++ partitionSchema.fields)

else
val resultSchemaString =  OrcUtils.orcTypeDescriptionString(StructType(requiredSchema.fields ++ partitionSchema.fields)))

since we are using this resultSchemaString in

batchReader.initBatch(
        TypeDescription.fromString(resultSchemaString),
        resultSchema.fields,

cloud-fan replied:
adding a new helper method is also good.

SaurabhChawla100 replied:
Added the helper method

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala:
```diff
@@ -288,4 +288,33 @@ class HiveOrcQuerySuite extends OrcQueryTest with TestHiveSingleton {
     }
   }
+
+  test("SPARK-32234: orc data created by the hive tables having _col fields name" +
```
we can shorten the test name: `SPARK-32234: read ORC table with column names all starting with '_col'`

SaurabhChawla100 replied:
done

cloud-fan reviewed (approved):
LGTM except some code style issues. Thanks for the fix!

SaurabhChawla100 committed:
created new method for duplicate code

SparkQA commented:
**[Test build #125892 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125892/testReport)** for PR 29045 at commit [`cf68729`](https://github.com/apache/spark/commit/cf6872989fdcb5396357c0e4cd3b3529e1334e6a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30500/
Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcPartitionReaderFactory.scala:
```diff
@@ -112,25 +112,26 @@ case class OrcPartitionReaderFactory(
+      val (requestedColIds, canPruneCols) = resultedColPruneInfo.get
+      val resultSchemaString = OrcUtils.orcResultSchemaString(canPruneCols,
+        dataSchema, resultSchema, partitionSchema, conf)
+      val requestedDataColIds = requestedColIds ++ Array.fill(partitionSchema.length)(-1)
```
I think we should switch the name. Here we add the partition column IDs and better to call it `requestedColIds`. The former one can be called `requestedDataColIds` as it doesn't contain partition columns.

SaurabhChawla100 replied:
switched the name

SaurabhChawla100 committed:
switch the name of requestedColIds and requestedDataColIds

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30515/
Test PASSed.

SparkQA commented:
**[Test build #125907 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125907/testReport)** for PR 29045 at commit [`6150b08`](https://github.com/apache/spark/commit/6150b08470384aa2280c5ffd141971aa6ff52410).

SparkQA commented:
**[Test build #125892 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125892/testReport)** for PR 29045 at commit [`cf68729`](https://github.com/apache/spark/commit/cf6872989fdcb5396357c0e4cd3b3529e1334e6a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125892/
Test FAILed.

SparkQA commented:
**[Test build #125907 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125907/testReport)** for PR 29045 at commit [`6150b08`](https://github.com/apache/spark/commit/6150b08470384aa2280c5ffd141971aa6ff52410).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125907/
Test FAILed.

SaurabhChawla100 committed:
fix the name of the unit test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30564/
Test PASSed.

SparkQA commented:
**[Test build #125949 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125949/testReport)** for PR 29045 at commit [`c0f6209`](https://github.com/apache/spark/commit/c0f62095f272aed48f0055afe8e7777a2f90fc54).

SparkQA commented:
**[Test build #125949 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/125949/testReport)** for PR 29045 at commit [`c0f6209`](https://github.com/apache/spark/commit/c0f62095f272aed48f0055afe8e7777a2f90fc54).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/125949/
Test FAILed.

SaurabhChawla100 commented:
@cloud-fan / @dongjoon-hyun  - The test build is failing with following error 
```
Test build #125949 has finished for PR 29045 at commit c0f6209.

This patch fails PySpark pip packaging tests.
This patch merges cleanly.
This patch adds no public classes.
```
I am not sure if its related to change done on this PR

cloud-fan commented:
All the github action checks passed, I think it's good to go. Merging to master/3.0!

dongjoon-hyun commented:
Thank you, @SaurabhChawla100 and @cloud-fan .

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:
```diff
@@ -199,4 +205,25 @@ object OrcUtils extends Logging {
+   * @return Returns the result schema string based on the canPruneCols flag.
+   *         resultSchemaString will be created using resultsSchema in case of
+   *         canPruneCols is true and for canPruneCols as false value
+   *         resultSchemaString will be created using the actual dataSchema.
```
This description is not clear enough. This utility function also changed the value of `conf`. We need to document it. 

@SaurabhChawla100 Could you submit a follow-up PR to improve the description?

SaurabhChawla100 replied:
@gatorsmile - This is the new helper method that we have added as the part of this PR

sure I Will update the description of PR . Shall I raised the PR against the new Jira or with this same jira . Since this Jira is already resolved

maropu replied:
> Shall I raised the PR against the new Jira or with this same jira

Its okay to refer to this JIRA ticket. Then, please add `[FOLLOWUP]` in the PR title.
